padsearch/padsearch.py

```diff
@@ -1,560 +1,626 @@
-    bot.add_cog(n)
+import json
+import math
+
+import discord
+from discord.ext import commands
+from ply import lex, yacc
+from png import itertools
+
+from __main__ import user_allowed, send_cmd_help
+
+from . import rpadutils
+from .utils import checks
+from .utils.chat_formatting import box, inline
+
+HELP_MSG = """
+^search <specification string>
+
+Colors can be any of:
+  fire water wood light dark
+Additionally, orb colors can be any of:
+  heart jammer poison mortal
+
+Options which take multiple colors should be comma-separated.
+
+Single instance filters
+* cd(n)       : Min cd <= n
+* farmable    : Obtainable outside REM
+* haste(n)    : Skill cds reduced by n
+* inheritable : Can be inherited
+* shuffle     : Board shuffle (aka refresh)
+* unlock      : Orb unlock
+* delay(n)    : Delay enemies by n
+* attabsorb   : Attribute Absorb shield null
+* absorbnull  : Damage Abasorb shield null
+* combo(n)    : Increase combo count by n
+* shield(n)   : Reduce damage taken by n%
+
+Multiple instance filters 
+* active(str)     : Active skill name/description
+* board(colors,)  : Board change to a comma-sep list of colors
+* color(color)    : Primary monster color
+* column(color)   : Creates a column of a color
+* hascolor(color) : Primary or secondary monster color
+* leader(str)     : Leader skill description
+* name(str)       : Monster name 
+* row(color)      : Creates a row of a color
+* type(str)       : Monster type
+* convert(c1, c2) : Convert from color 1 to color 2, accepts any as entry as well
+"""
+
+COLORS = [
+    'fire',
+    'water',
+    'wood',
+    'light',
+    'dark',
+]
+
+TYPES = [
+    "attacker",
+    "awoken",
+    "balance",
+    "devil",
+    "dragon",
+    "enhance",
+    "evolve",
+    "god",
+    "healer",
+    "machine",
+    "physical",
+    "protected",
+    "vendor",
+]
+
+ORB_TYPES = [
+    'any',
+    'fire',
+    'water',
+    'wood',
+    'light',
+    'dark',
+    'heal',
+    'jammer',
+    'poison',
+    'mortal',
+]
+
+
+def assert_color(value):
+    value = replace_named_color(value)
+    if value not in COLORS:
+        raise rpadutils.ReportableError(
+            'Unexpected color {}, expected one of {}'.format(value, COLORS))
+    return value
+
+
+def assert_orbcolor(value):
+    value = replace_named_color(value)
+    if value not in ORB_TYPES:
+        raise rpadutils.ReportableError(
+            'Unexpected orb {}, expected one of {}'.format(value, ORB_TYPES))
+    return value
+
+
+def assert_orbcolors(values):
+    return [assert_orbcolor(c) for c in values]
+
+
+def split_csv_orbcolors(value):
+    parts = [p.strip() for p in value.split(',')]
+    return assert_orbcolors(parts)
+
+
+COLOR_REPLACEMENTS = {
+    'any': '',
+    'red': 'fire',
+    'r': 'fire',
+    'blue': 'water',
+    'b': 'water',
+    'green': 'wood',
+    'g': 'wood',
+    'l': 'light',
+    'd': 'dark',
+    'heart': 'heal',
+    'h': 'heal',
+    'p': 'poison',
+    'mp': 'mortalpoison',
```

Comment:
Instead of using two letters for these, why not just m for mortal and o for bomb (that's what I changed dawnglare to do).